File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;
import java.io.*;

class Solution {
    static int answer = 0;
    public int solution(int[] numbers, int target) {
        //숫자배열, 노드 깊이, 타겟넘버, 이전까지의 합
        DFS(numbers,0,target,0);
        
        return answer;
    }
    
    public static void DFS(int[] numbers, int depth, int target, int sum){
        //마지막 노드까지 탐색한 경우
        if(depth == numbers.length){
            if(sum == target) answer++;
            return;
        }
        
        DFS(numbers, depth+1, target, sum + numbers[depth]);
        DFS(numbers, depth+1, target, sum - numbers[depth]);
    }
    
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    answer = 0
    
    def dfs(idx, number_sum):
        nonlocal answer
        
        if idx == len(numbers):
            if number_sum == target:
                answer += 1
            return
            
        dfs(idx+1, number_sum + numbers[idx])
        dfs(idx+1, number_sum - numbers[idx])
        
    
    dfs(0,0)
    
    return answer